core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/pythonUDF/PythonUDFOpExec.java

```diff
@@ -36,75 +35,23 @@
-     * {@link FlightClient#startPut(FlightDescriptor, VectorSchemaRoot, FlightClient.PutListener, CallOption...)} and
-     * {@link FlightClient.ClientStreamListener#putNext()}. The server uses {@code do_put()} to receive data stream
-     * and convert it into a {@code pyarrow.Table} and store it in the server.
-     * {@code startPut} is a non-blocking call, but this method in general is a blocking call, it waits until all the
-     * data are sent.
-     *
-     * @param client      The FlightClient that manages this.
-     * @param values      The input queue that holds tuples, its contents will be consumed in this method.
-     * @param arrowSchema Input Arrow table schema. This should already have been defined (converted).
-     * @param channel     The predefined path that specifies where to store the data in Flight Serve.
-     * @param chunkSize   The chunk size of the arrow stream. This is different than the batch size of the operator,
-     *                    although they may seem similar. This doesn't actually affect serialization speed that much,
-     *                    so in general it can be the same as {@code batchSize}.
-     */
-    private void writeArrowStream(FlightClient client, Queue<Tuple> values,
-                                  org.apache.arrow.vector.types.pojo.Schema arrowSchema,
-                                  Channel channel, int chunkSize) throws RuntimeException {
-        SyncPutListener flightListener = new SyncPutListener();
-        VectorSchemaRoot schemaRoot = VectorSchemaRoot.create(arrowSchema, PythonUDFOpExec.memoryAllocator);
-        FlightClient.ClientStreamListener streamWriter = client.startPut(FlightDescriptor.path(Collections.singletonList(channel.name)), schemaRoot, flightListener);
-        try {
-            while (!values.isEmpty()) {
-                schemaRoot.allocateNew();
-                int indexThisChunk = 0;
-                while (indexThisChunk < chunkSize && !values.isEmpty()) {
-                    convertAmber2ArrowTuple(values.remove(), indexThisChunk, schemaRoot);
-                    indexThisChunk++;
-                }
-                schemaRoot.setRowCount(indexThisChunk);
-                streamWriter.putNext();
-                schemaRoot.clear();
-            }
-            streamWriter.completed();
-            flightListener.getResult();
-            flightListener.close();
-            schemaRoot.clear();
-        } catch (Exception e) {
-            closeAndThrow(client, e);
-        }
-    }
-
-    private String pythonScriptPath;
     private final String pythonScriptText;
     private final ArrayList<String> inputColumns;
     private final ArrayList<Attribute> outputColumns;
-
     private final ArrayList<String> arguments;
     private final ArrayList<String> outerFilePaths;
     private final int batchSize;
     private final boolean isDynamic;
-
+    private final Queue<Tuple> inputTupleBuffer = new LinkedList<>();
```

Comment:
this entire file is heavily changed. Are all these changes relevant to logging? Or is this a problem with git diff renderer?

Reply:
Most of the differences are due to reformat.

Significant changes:
1. The way we wrap exceptions as JSON output is no longer used. Now PyArrow server reports user exceptions, with the RPC protocal.
2. Java PyArrow client can handle any exceptions by forwarding it to the frontend.